dbms/Storages/MergeTree/KeyCondition.h

```diff
@@ -26,17 +26,61 @@ using FunctionBasePtr = std::shared_ptr<IFunctionBase>;
 class ExpressionActions;
 using ExpressionActionsPtr = std::shared_ptr<ExpressionActions>;
 
+/** A field, that can be stored in two reperesenation:
+  * - A standalone field.
+  * - A field with reference to it's position in block.
+  *   It's needed for execution functions on ranges during
+  *   index analysis. If function was executed once for field,
+  *   it's result would be cached for all block for which field's reference points to.
+  */
+struct FieldRef
+{
+public:
+    using SharedBlock = std::shared_ptr<Block>;
+
+    FieldRef() = default;
+
+    /// Create as explicit field without block.
+    template <typename T>
+    FieldRef(const T & field) : cached_field(field) {}
+
+    /// Create as reference to field in block.
+    FieldRef(const SharedBlock & block_, size_t row_idx_, size_t column_idx_)
+        : block(block_), row_idx(row_idx_), column_idx(column_idx_) {}
+
+    Field * operator->() { return &get(); }
+    const Field * operator->() const { return &get(); }
+    const Field & operator*() const { return get(); }
+
+    bool isExplicit() const { return block == nullptr && !cached_field.isNull(); }
+
+    SharedBlock block;
+    size_t row_idx;
+    size_t column_idx;
+
+private:
+    Field & get() const
+    {
+        if (cached_field.isNull() && block)
+            cached_field = (*block->getByPosition(column_idx).column)[row_idx];
+
+        return cached_field;
+    }
+
+    mutable Field cached_field;
```

Comment:
It looks like you can just inherit `Field` instead of having a wrapper class and without all the `->` hassles

Reply:
Yes, it's more convinient.